doc/manual/typing.md

```diff
@@ -0,0 +1,629 @@
+wanted: the argument must have a field `total: Num`, but the *tail* (the rest of
+the record type) is polymorphic, and `a` may be substituted for arbitrary fields
+(such as `jan: Num, feb: Num`). We used two different generic parameters `a` and
+`b`, to express that the tails of the arguments may differ.  If we used `a` in
+both places, as in `forall a. {total: Num | a} -> {total: Num | a} -> Num`, we
+could still write `addTotal {total = 1, foo = 1} {total = 2, foo = 2}` but not
+`addTotal {total = 1, foo = 1} {total = 2, bar = 2}`. Using distinct parameters
+`a` and `b` gives us maximum flexibility.
+
+What comes before the tail may include several fields, is in e.g. `forall a.
+{total: Num, subtotal: Num | a} -> Num`.
+
+Note that row polymorphism also works with enums, with the same intuition of a
+tail that can be substituted for something else. For example:
+
+```nickel
+let portOf : forall a. <http, ftp | a> -> Num = fun protocol =>
+switch {
+  `http -> 80,
+  `ftp -> 21,
+  _ -> 8000,
+} protocol
+```
+
+Because the `switch` statement has a catch-all case `_`, this function is indeed
+able to handle other tags than `http` and `ftp`, as expressed by its polymorphic
+type.
+
+### Take-away
+
+The type system of Nickel has a few basic types (`Dyn`, `Num`, `Str`,
+and `Bool`) and type constructors for lists, records, enums and functions.
+Nickel features generics via polymorphism, introduced by the `forall` keyword. A
+type can not only be generic in other types, but records and enums types can
+also be generic in their tail. The tail is delimited by `|`.
+
+## Interaction between statically typed and dynamically typed code
+
+In the previous section, we've been focusing solely on the static typing side.
+We'll now explore how typed and untyped code interact.
+
+### Using statically typed code inside dynamically code
+
+Until now, we have written the statically typed `filter` examples using
+statically typed blocks that enclosed both the definition of `filter` and the
+call sites. More realistically, `filter` would be a statically typed library
+function (it is actually part of the standard library as `lists.filter`) while
+being likely to be called from dynamically typed configuration files. In this
+situation, the call site escapes the typechecker. Thus, without an additional
+mechanism, static typing would only ensure that the implementation of `filter`
+doesn't violate the typing rules, but wouldn't prevent a ill-formed call from
+dynamically typed code.  At first sight, we haven't solved at all the main issue
+of delayed dynamic type errors that static typing was supposed to solve in the
+first place! Remember, the typical problem is the caller passing an value of the
+wrong type that eventually raises an error from within `filter`.
+
+Hopefully, Nickel does have a mechanism to prevent this from happening and to
+provide good error reporting in this situation.  Let us see by ourselves:
```

Comment:
```suggestion
Fortunately, Nickel does have a mechanism to prevent this from happening and to
provide good error reporting in this situation.  Let us see by ourselves:
```